### DISCUSSION TITLE
Filter by indirect relation

### DISCUSSION THREAD

**Discussion by elliotsoomro** [pocketbase/pocketbase#1251] [Q&A] [answered]
Is is possible to use filtering based on an indirect relation?

E.g. filter: 'alternatives(question).text~<search_query>'

This example does not seem to work.

**Comment by ganigeorgiev** [comment, marked as answer]
No, this is not possible.

There is a `@collection.*` filter but this will work only as part of an API rule, eg:

```
@collection.alternatives.question = id && @collection.alternatives.question.text ~ 'abc'
```

It may be easier to just make 2 separate requests, something like.

```js
// fetch alternatives
const alternatives = await pb.collection('alternatives').getFullList(200, { filter: 'text ~ "abc"' });

// extract question ids in a single filter expression (eg. "id='1' || id='2' || ...")
const questionIdsExpr = alternatives.map((a) => `id='${a.question}'`).join('||');

// fetch questions with the above filter
if (questionIdsExpr) {
    const questions = await pb.collection('questions').getFullList(200, { filter: questionIdsExpr })
}
```

_**edit:** Or just reverse the search criteria, aka. get alternatives with expanded `questions`:_

```js
const alternativesWithQuestion = await pb.collection('alternatives').getFullList(200, {
    filter: 'text ~ "abc"',
    expand: 'question',
})
```

**Reply by elliotsoomro** [reply]
Is there a limit to how much data can be passed in a filter query param? I seem to get an 400 Bad Request error with message "Invalid filter parameters." when I pass a lot of question IDs. It works when I decrease the number of IDs. I can only pass around 70 maximum.

The filter query param looks like this:

`(((id="blynuki6g35pyz5"||id="xy4wh8x23e2s8qx"||id="lz1l9nrdsin07is"||id="0fpq6dcrit36us6"||id="f2abbxmv2lt6h7x"||id="0cxt3xpqe4q7rmj"||id="7qetbxe7v4jd6sc"||id="7qetbxe7v4jd6sc"||id="m674fnlvaceckpj"||id="m674fnlvaceckpj"||id="zvvmkmdowocrtaq"||id="8vb1umg99d1zn0t"||id="qkjj74661yb7rhu"||id="qkjj74661yb7rhu"||id="h1tpa46ktjcgncn"||id="h1tpa46ktjcgncn"||id="h1tpa46ktjcgncn"||id="a1qhjre1mkg7wg9"||id="s8hcb3bk2q7lfwk"||id="5dnl6v4cbrkbzcn"||id="9vs763d4x9is57j"||id="9vs763d4x9is57j"||id="xpunhyy1zbxtlt8"||id="sr703h0g903tcl0"||id="sr703h0g903tcl0"||id="gbkepy9avidrbmg"||id="gbkepy9avidrbmg"||id="gbkepy9avidrbmg"||id="ljss53chdj7yslf"||id="tr6zepu0ao9nlq9"||id="dh3vunzdwnscwf3"||id="iv8xzzpqiq4rmpd"||id="36mnb8opavkblnu"||id="36mnb8opavkblnu"||id="8u3udh6b1p0hb8k"||id="8u3udh6b1p0hb8k"||id="whznh9y7xzkmeus"||id="7jig5npaj8gnpck"||id="7jig5npaj8gnpck"||id="hxwzmimul18tai3"||id="hxwzmimul18tai3"||id="qu9akuo58adp5sl"||id="id6hpm3yww6hkie"||id="6166wx2wmdi9woc"||id="6166wx2wmdi9woc"||id="6166wx2wmdi9woc"||id="5ndavirt59guc6m"||id="b596qp5dzdryfcm"||id="b596qp5dzdryfcm"||id="5cya2bp6jv04zze"||id="49hdpf8jpcnsc3y"||id="49hdpf8jpcnsc3y"||id="49hdpf8jpcnsc3y"||id="49hdpf8jpcnsc3y"||id="17x8cvpelpdsrs0"||id="17x8cvpelpdsrs0"||id="17x8cvpelpdsrs0"||id="d11yet57e72suem"||id="d11yet57e72suem"||id="gggnze88rqmuo12"||id="gggnze88rqmuo12"||id="vdg6hei41pczhht"||id="vdg6hei41pczhht"||id="h30cfhk7w97pwic"||id="7zol1sjjrfanrzc"||id="00xdw42unb81yno"||id="uk9zqkwivdd3idk"||id="uk9zqkwivdd3idk"||id="q9gg10q8rwhjh6n"||id="i8jhy9kl8ev8aar"||id="t2qt6eew5iidj08"||id="t2qt6eew5iidj08"||id="vl7d99duy8hzzay"||id="vl7d99duy8hzzay"||id="zkkrhevh3n23e4r"||id="zkkrhevh3n23e4r"||id="cgvloo587tcw0rh"||id="of3shpm5z60r56i"||id="zksdhogw98x5958"||id="zksdhogw98x5958")||text~"<search_query>"))`

**Reply by ganigeorgiev** [reply]
@elliotsoomrots It is possible that you are hitting the max length of an sql statement. Could you start PocketBase with `--debug` and/or check the more detailed error message in Admin UI > Logs?

**Reply by ganigeorgiev** [reply]
~~_Also note that in the above filter it seems that there is one `(` character more, so it could also be the reason why the filter expression fails._~~ My mistake, after double checking the parenthesis seems correct.

**Reply by elliotsoomro** [reply]
Using `--debug` I get the error `SQL logic error: parser stack overflow (1)`

**Reply by ganigeorgiev** [reply]
@elliotsoomro What version of PocketBase are you using?

**Reply by elliotsoomro** [reply]
v0.9.2

**Reply by ganigeorgiev** [reply]
Is it possible to post here the failing query (you can get it from the console right before the failure or if not available try with smaller number of ids so that it can print the sql statement)? I suspect that the length of the query string is causing it to fail but it could also be the parenthesis depth.

**Reply by elliotsoomro** [reply]
`[0.00ms] SELECT {{_collections}}.* FROM '_collections' WHERE [[id]] = 'questions' OR LOWER([[name]])='questions' LIMIT 1`
`[0.00ms] SELECT {{_collections}}.* FROM '_collections' WHERE [[id]] = 'poc519muhq7awnn' OR LOWER([[name]])='poc519muhq7awnn' LIMIT 1`
`[0.00ms] SELECT {{_collections}}.* FROM '_collections' WHERE [[id]] = 'r25nx2fzw4ccqh5' OR LOWER([[name]])='r25nx2fzw4ccqh5' LIMIT 1`
`[0.00ms] SELECT {{_collections}}.* FROM '_collections' WHERE [[id]] = 'systemprofiles0' OR LOWER([[name]])='systemprofiles0' LIMIT 1`
`[0.00ms] SELECT COUNT(*) FROM (SELECT DISTINCT 'questions'.'id' FROM 'questions' LEFT JOIN json_each(CASE WHEN json_valid([[questions.collection]]) THEN [[questions.collection]] ELSE json_array([[questions.collection]]) END) 'questions_collection_je' LEFT JOIN 'collections' 'questions_collection' ON [[questions_collection.id]] = [[questions_collection_je.value]] LEFT JOIN json_each(CASE WHEN json_valid([[questions_collection.folder]]) THEN [[questions_collection.folder]] ELSE json_array([[questions_collection.folder]]) END) 'questions_collection_folder_je' LEFT JOIN 'folders' 'questions_collection_folder' ON [[questions_collection_folder.id]] = [[questions_collection_folder_je.value]] LEFT JOIN json_each(CASE WHEN json_valid([[questions_collection_folder.user]]) THEN [[questions_collection_folder.user]] ELSE json_array([[questions_collection_folder.user]]) END) 'questions_collection_folder_user_je' LEFT JOIN 'users' 'questions_collection_folder_user' ON [[questions_collection_folder_user.id]] = [[questions_collection_folder_user_je.value]] WHERE (COALESCE('tai48wezcjtdkk3', '') = COALESCE([[questions_collection_folder_user.id]], '')) AND (((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((((COALESCE([[questions.id]], '') = COALESCE('ug0l7pdau5km56f', '')) OR (COALESCE([[questions.id]], '') = COALESCE('neqj8s9ma84d45s', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('dkbl72v2xjcdlyk', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('gub7vhgra83ygdd', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('wuuqibpg9taia0h', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('jma7ka76epd4k9t', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('eln1bw0n6f675ac', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('wh1dezujq4f2efl', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('k2qem0fekewt752', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('4oisune47iwuzx5', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('y7ubay2p27yd9ai', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('7i7msnhpjnyzctg', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('ryahi8erb6lu7jp', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('2gn2xci7s2ogfyo', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('rnkjor6ajy6quab', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('tt93o2axhrovbhi', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('4yduz9ka43sbwkr', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('bw86xc3prhjis2e', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('f43jjqpxq9fcvpk', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('y1ielh5k76zwemg', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('v5gsc3857ekx4g6', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('rx015cd9t7edz43', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('zc5x7mxyyhh0voc', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('btwnlr6twldauc4', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('hrgfiupcbty8c6t', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('5q9g0yszaiuzlk1', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('12sfu450tb8y017', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('m64ndsh2a1d4qs0', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('0w9w4325gbvhkl9', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('xvzh1i1w3vg65i3', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('0f3vhpc4bsewa6b', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('4ex9hu7dbu9q8uo', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('3to4enpa50f8rch', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('nhcyxx6h3ulmotu', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('ea18x6shra8cysn', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('v6ji8mqbqz1e6ij', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('petdwmtt5crfi9o', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('vsb5vss9y137cdc', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('ibmkcyhqkaym6nw', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('ugvbvcmhvby9q6o', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('6tledk5qae0yk92', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('s3w4rbt071xtror', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('q3ok5nejidm2pre', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('56mw0wk3zkhq1ru', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('m2bk8naj3ytg2up', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('nnar2gf18pe0unu', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('0fk7x3pphrfurpg', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('2f2z3o0pg7s40xo', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('xg21sr3hp53wg5m', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('gsqdro8k625499d', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('cnscv40zu66dvac', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('xodjep0f1usfdve', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('isxnfuka2mf04yz', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('u72vq6wwu9np7qx', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('puv58a09n57lwt1', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('flvop7k75omuwdn', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('iqhhyenzqppl8rk', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('qc1dcucl4ww51v1', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('b56evgc0mn4ke9c', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('6h8lj5j72zup7uz', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('qfugviofvc7ofx8', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('x2vkpjkcw9m28xi', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('aih49m3xio4jtra', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('ip7gdnttknw84e0', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('78vawwf3x86t1nw', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('pg69k2je91tmisl', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('i0y6t77upnbtjv8', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('9x20sk62668opyo', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('oddt8wmup826i6b', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('vzxbl28cf4gtrjn', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('ksmy2aum5g4iie3', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('fucec2krwpsieqg', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('olna97s566mvbvt', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('i73wihtnx209bz7', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('y0h3g4th8wwl443', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('omqu5uzszjfndbb', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('hxewcl5vfpsahfz', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('5ynq3bi3so73q44', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('te0wh4p3gzoxcdv', ''))) OR (COALESCE([[questions.id]], '') = COALESCE('mlo99n9beup702a', ''))) OR ([[questions.text]] LIKE '%a%')))`
`2022/12/13 19:07:40 SQL logic error: parser stack overflow (1)`

**Reply by ganigeorgiev** [reply]
Thank you. It is probably the parenthesis that is causing it to fail. I'm currently working on some filter enhancements and will try to optimize the generated query string for the next release.

As a temporary workaround you can try to use the second approach (aka. fetching the `alternatives` each with their expanded question) OR batch the ids, eg. something like:

```js
let questions = [];
let questionIds = alternatives.map((a) => a.question);

while (questionIds.length) {
    const idsExpr = questionIds.splice(0, 50).map((id) => `id='${id}'`).join('||');
    questions = questions.concat(await pb.collection('questions').getFullList(200, { filter: idsExpr }));
}
```

**Comment by Sid-Turner-Ellis** [comment]
@ganigeorgiev in the example you gave (`@collection.alternatives.question = id && @collection.alternatives.question.text ~ 'abc'`), will  @collection.alternatives.question.text be filtered from the first filter? 

I have a database:

- users

- projects
  - owners (many users)  
 
- tasks
  - project (single project)

And i'm trying to create a create rule for requests which translates to: "The requests user ID must be present in the users array of the project your trying to create the task for"

I would have thought his would work but it's not doing what I expected:

`@collection.projects.id ?= @request.data.project.id && @collection.projects.users ?=  @request.auth.id`

**Reply by ganigeorgiev** [reply]
The above discussion is very old and I'm not sure if I'm understanding correctly your collections structure, but you can try with a "tasks" rule like:

```js
@request.auth.id != "" &&
project.owners.id ?= @request.auth.id
```

If the above doesn't help, please file a new discussion with more details about db structure and for which collection you want to apply the API rule.

**Reply by Sid-Turner-Ellis** [reply]
Thanks, i've created a discussion as that one isn't working for me - I initially thought that would work but assumed it wasn't working because it was in the 'create' rule and it wouldn't have access to the project yet 

https://github.com/pocketbase/pocketbase/discussions/3523
